### PR TITLE
feat: add a ping tool

### PR DESCRIPTION
**Pull request by marcelo-gonzalez**
This connects to a node, and tries to send `Ping` messages to the targets we get in the `AnnounceAccount`s received from that node, keeping track of latencies and timeouts. To use it:

```
./target/release/neard ping --chain-id {some chain id} --peer "ed25519:{my node's public key}@{my node's socket addr}"
```

and you should get something like:

```
send ping --------------> foo.pool.near
recv pong <-------------- foo.pool.near latency: 867.94628ms
send ping --------------> bar.pool.near
send ping --------------> baz.pool.near
recv pong <-------------- bar.pool.near latency: 1.042379472s
recv pong <-------------- baz.pool.near latency: 183.646969ms
...
...
```

### CODE REVIEW

**Comment by marcelo-gonzalez** [comment]
this is basically plus some extra fixes/cleanups

@pompon0 feel free to skip reviewing the whole thing, since I think @nikurt has looked at a lot of the code already. I have not forgotten your earlier comment :) (https://github.com/near/nearcore/pull/7303) and my response to it saying that it sounded like the way to go.  But after looking more at doing it that way to actually merge this, it looks to me like if we want to use the current `near-network` code to implement this, we need a `PeerManagerActor` and a `PeerActor` for the node we're connecting to, since what we want here is `PeerManagerActor::send_ping()`/`PeerManagerActor::send_signed_message_to_peer()`. And to construct these, we'd need a `near_store::Store`, plus a `ClientActor` and `ViewClientActor`.  That is, we need the whole aparatus that comes with running a full node. This doesn't sound so bad, but I think it leads to some practical problems:


1. Using near-network would require a valid data dir for the chain whose network we're trying to send pings to, which makes things a lot more cumbersome. Currently this tool lets you do:

```
ping --chain-id 'mainnet' --peer ed25519:abc@1.2.3.4:24567
```

wait for a bit to collect some stats, and then hit ^C and right away run it on a different chain:

```
ping --chain-id 'testnet' --peer ed25519:xyz@5.6.7.8:24567
```

And all this can be done on a small laptop with no near home directory downloaded. If we use the near-network package, then we need a valid data dir.

2. The current logic of connecting to exactly 1 peer specified on the cmd line (or maybe in the future several peers, but still I think we'd want to specify them directly) and sending pings through it is awkward to acheive with the peer manager. Unless we make significant changes (which feel too invasive to do just for this), we need to set several different network config options to 1 (max_num_peers for example) and we need to make sure the peer store only has the peer we want to connect to in it, which is not easy when the peer manager keeps adding peers it hears about after sending a `PeersRequest` (as it should)

3. Like in my comment I linked above, it's kinda nice to be able to measure latencies right after reading from the TCP stream instead of after receiving actix messages. Kinda gives more trust in the numbers. And in the code in this PR, I've tried to make that an actual priority, where the recv_messages() function reads as many messages as are available in the buffers before deserializing them, so we can take timestamps ASAP (the read_remaining_messages() function here just returns a Vec of lengths and timestamps, then we deserialize after that)

4. The peer manager is gonna be doing all kinds of stuff that we do not need. What we're doing here is really just pretending to be a legitimate implementor of the NEAR protocol (we send a handshake after all), but then we are actually lying, and we do nothing but send pings. We even have flags like --protocol-version, --head-height, --genesis-hash, etc to change values in our handshake on the fly. So since we're not actually implementing the protocol, we don't really need, for example, a `RoutingTableActor` running in the background, and other stuff like that.

TLDR: if we just reimplement the basic wire protocol stuff we can avoid needing a home dir, the logic is easier to implement, and the quality of latency numbers is better.

anyway WDYT? If I have missing/not considered something here, happy to do it another way.

**Comment by pompon0** [comment]
@marcelo-gonzalez you don't need the whole node, spawning just a PeerActor (or perhaps a PeerManagerActor) is enough (with faked everything else). Please take a look at the tests in the near-network package. And even if you don't want to do that, that's ok, but please do not make the guts of the near-network package public. Instead add whatever you need to the near-network package with exposes a concise and precise API: just one public function near_network::ping_node() would be ok, exposing random parts of near-network is not ok.

**Review by pompon0 — CHANGES_REQUESTED** [review]
(see prev comment)

File: tools/ping/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,734 @@
...
+                "no more bytes available, but expected to receive a message",
+            ));
+        }
+        Ok(())
+    }
+
+    // read at least 4 bytes, but probably the whole message in most cases
+    async fn read_msg_length(&mut self) -> io::Result<(usize, Instant)> {
+        let mut first_byte_time = None;
+        while self.buf.remaining() < 4 {
```
Comment (nikurt) [inline tools/ping/src/lib.rs:112]: Please consider using `size_of` instead of a magic constant `4`.

**Review by nikurt — APPROVED** [review]

**Comment by matklad** [comment]
Mild preference for adding this as an subcommand of `neard` rather than as a standalone tool: 

* each binary significantly increases build time (build time is approximately quadratic in the number of binaries)
* it's useful to have all debugging tools to be built-in into the node, such that, when you want to debug something in the field, you can just use the `neard` binary.

**Comment by marcelo-gonzalez** [comment]
> @marcelo-gonzalez you don't need the whole node, spawning just a PeerActor (or perhaps a PeerManagerActor) is enough (with faked everything else). Please take a look at the tests in the near-network package. And even if you don't want to do that, that's ok, but please do not make the guts of the near-network package public. Instead add whatever you need to the near-network package with a concise and precise API: just one public function near_network::ping_node() would be ok, exposing random parts of near-network is not ok.

ok fair enough. Feels to me like mocking the inputs to `PeerManagerActor::new()` to achieve simpler logic is quite indirect/more complicated than it needs to be. @pompon0  how do you feel about adding something like `pub mod raw_peer` (or maybe with a better name than that) to near-network, with the `Peer` struct in this PR, and these pub functions:


```rust
pub struct Peer {
    stream: TcpStream,
    peer_id: PeerId,
    buf: BytesMut,
}

impl Peer {
    pub async fn connect(addr: SocketAddr, peer_id: PeerId) -> io::Result<Self> {
    }

    pub async fn recv_messages(&mut self) -> anyhow::Result<Vec<(PeerMessage, Instant)>> {
    }

    pub async fn write_message(&mut self, msg: &PeerMessage) -> io::Result<()> {
    }
}
```

then plus side is any other code that wants to connect to a peer but not with a "real"/full node can use this. One other potential user I can think of is the `mock-node` code that currently mocks stuff at the level of the `ClientActor`, but w/ something like the above could be made to mock the whole thing

**Comment by pompon0** [comment]
@marcelo-gonzalez yes, that's closer to what I have in mind. Still I'd prefer the PeerMessage to stay private to the module, so instead of generic `send/recv_message`,
provide a separate method for each use case. In your case it can be `pub async fn ping()` which sends the ping and then waits for the response.

**Comment by marcelo-gonzalez** [comment]
> @marcelo-gonzalez yes, that's closer to what I have in mind. Still I'd prefer the PeerMessage to stay private to the module, so instead of generic `send/recv_message`, provide a separate method for each use case. In your case it can be `pub async fn ping()` which sends the ping and then waits for the response.

@pompon0 ah but looks like `PeerMessage` is already `pub` right? Do you mean just that its direct use should be discouraged anyway, or you have plans to remove that `pub`? Yea in that case maybe I can change the above to:

```rust
pub struct Peer {
    stream: TcpStream,
    peer_id: PeerId,
    buf: BytesMut,
}

pub enum Message {
    Pong(...)
}

impl Peer {
    pub async fn connect(addr: SocketAddr, peer_id: PeerId) -> io::Result<Self> {
    }

    pub async fn recv_messages(&mut self) -> anyhow::Result<Vec<(Message, Instant)>> {
    }

    pub async fn send_ping(&mut self, target: AccountId) -> io::Result<()> {
    }
}
```

where `recv_messages` will drop anything other than a `Pong`, and in the future we can add stuff to the `Messages` enum (like in the `mock-node` use case I mentioned, maybe we wanna receive `BlockHeadersRequest`s for example). Because I think the problem with having `pub async fn ping()` also wait for the response is that we're sending pings to a bunch of different targets and we don't want to wait for each one to get back to us before we send the next one. Of course could just spawn all the futures but feels much more direct/less overhead to have send_ping() resolve after the message is sent.

**Comment by marcelo-gonzalez** [comment]
> Mild preference for adding this as an subcommand of `neard` rather than as a standalone tool:
> 
> * each binary significantly increases build time (build time is approximately quadratic in the number of binaries)
> * it's useful to have all debugging tools to be built-in into the node, such that, when you want to debug something in the field, you can just use the `neard` binary.

oof, was not aware of that first point. But I wonder... if we include this in the `neard` binary, maybe people will say "ooh, what is this ping command??" and run it when they otherwise would have had no intention of doing that. Of course it's always possible for people to send pings at will, but I wonder if adding this to neard would lead to a nontrivial increase in useless messages being routed thru the network as people use the tool casually. Prob just being paranoid though, will update the PR to make it a neard subcommand in a bit

**Comment by pompon0** [comment]
> ah but looks like PeerMessage is already pub right? Do you mean just that its direct use should be discouraged anyway, or you have plans to remove that pub? 

Yes, PeerMessage should be private, but unfortunately some integration tests which live in a different crate have to have access to it. Once I migrate all those tests to near-network, PeerMessage will be pub(crate). The direct usage is discouraged. Sorry, I wasn't clear on that one.

> Of course could just spawn all the futures but feels much more direct/less overhead to have send_ping() resolve after the message is sent.

Well, that's my ultimate goal: to make network API just a bunch of RPCs for which you can await the response directly. Here is an old prototype of mine, which implements a response-dispatching event loop (on top of PeerManager): https://github.com/near/nearcore/blob/947af6b0d4c4c892ba5083f34254fdf89188e57f/tools/chainsync-loadtest/src/network.rs . But yeah, getting it production ready will require some work, so the API that you suggested (with `pub enum Message { Pong(...) }`) should be fine for now.

**Comment by marcelo-gonzalez** [comment]
> Well, that's my ultimate goal: to make network API just a bunch of RPCs for which you can await the response directly. 

I see, ok I'll update it to be done that way. Maybe next week if time permits

**Comment by marcelo-gonzalez** [comment]
ok finally got around to this, and updated it to add a new `pub mod raw` to `near_network`, and to make it a subcommand of `neard`. @pompon0 PTAL

File: chain/network/src/raw.rs [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+    my_peer_id: PeerId,
+    peer_id: PeerId,
+    secret_key: SecretKey,
+    stream: TcpStream,
+    buf: BytesMut,
+}
+
+/// The types of messages it's possible to receive from a `Peer`. Any PeerMessage
+/// we receive that doesn't fit one of these will just be logged and dropped.
+pub enum Message {
```
Comment (pompon0) [inline chain/network/src/raw.rs:35]: if this enum is specifically representing received messaged, then rename it to sth like "Response".

File: chain/network/src/raw.rs [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+    AnnounceAccounts(Vec<(AccountId, PeerId, EpochId)>),
+    Pong { nonce: u64, source: PeerId },
+}
+
+impl Message {
+    fn from_peer_message(m: PeerMessage) -> Option<Self> {
```
Comment (pompon0) [inline chain/network/src/raw.rs:41]: how about making it an implementation of TryFrom ?

File: chain/network/src/raw.rs [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+use std::time::{Duration, Instant};
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::TcpStream;
+
+/// Represents a connection to a peer, and provides only minimal functionality.
+/// Almost none of the usual NEAR network logic is implemented, and the caller
+/// will receive messages via the recv() function. Currently the only message
+/// you can send is a Ping message, but in the future we can add more stuff there
+/// (e.g. sending blocks/chunk parts/etc)
+pub struct Peer {
```
Comment (pompon0) [inline chain/network/src/raw.rs:25]: Rename it to Connection, since it is a "connection to a peer"

File: chain/network/src/raw.rs [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+            Err(e) => format!("Err({:?})", e),
+        };
+        write!(f, "{:?}@{:?}", &self.peer_id, addr)
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ConnectError {
+    #[error(transparent)]
+    IO(#[from] std::io::Error),
```
Comment (pompon0) [inline chain/network/src/raw.rs:70]: nit: prefer `#[source]` to `#[from]` to avoid obfuscation of error conversion.

File: chain/network/src/raw.rs [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+use bytes::BytesMut;
+use near_crypto::{KeyType, SecretKey};
+use near_primitives::block::GenesisId;
+use near_primitives::hash::CryptoHash;
+use near_primitives::network::PeerId;
+use near_primitives::types::{AccountId, BlockHeight, EpochId};
+use near_primitives::version::{ProtocolVersion, PROTOCOL_VERSION};
+use std::io;
+use std::net::SocketAddr;
+use std::time::{Duration, Instant};
```
Comment (pompon0) [inline chain/network/src/raw.rs:16]: use crate::time instead. It will incur some extra conversions when interacting with tokio API, but that's the direction we are aiming at.

File: chain/network/src/raw.rs [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+    ) -> Result<Self, ConnectError> {
+        let secret_key = SecretKey::from_random(KeyType::ED25519);
+        let my_peer_id = PeerId::new(secret_key.public_key());
+
+        let start = Instant::now();
+        let stream = tokio::time::timeout(Duration::from_secs(2), TcpStream::connect(addr))
+            .await
+            .map_err(Into::<io::Error>::into)??;
+        tracing::info!(
+            target: "network", "Connection to {:?}@{:?} established. latency: {:?}",
```
Comment (pompon0) [inline chain/network/src/raw.rs:94]: {peer_id}@{addr:?}

File: chain/network/src/raw.rs [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+            // TODO: maybe check the handshake for sanity
+            PeerMessage::Handshake(_) => {
+                tracing::info!(target: "network", "handshake latency: {:?}", timestamp - start);
+            }
+            PeerMessage::HandshakeFailure(_peer_info, reason) => {
+                return Err(ConnectError::HandshakeFailure(reason))
+            }
+            _ => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
```
Comment (pompon0) [inline chain/network/src/raw.rs:156]: technically this is not a transport layer error, so perhaps add a dedicated variant to ConnectError, sth like UnexpectedMessage?

File: chain/network/src/raw.rs [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+    }
+
+    async fn write_message(&mut self, msg: &PeerMessage) -> io::Result<()> {
+        let mut msg = msg.serialize(Encoding::Proto);
+        let mut buf = (msg.len() as u32).to_le_bytes().to_vec();
+        buf.append(&mut msg);
+        self.stream.write_all(&buf).await
+    }
+
+    // panics if there's not at least `len` + 4 bytes available
```
Comment (pompon0) [inline chain/network/src/raw.rs:176]: why not make it async and wait for the expected number of bytes to arrive?

File: chain/network/src/raw.rs [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+        msg.map_err(|_| {
+            io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!("error parsing protobuf of length {}", len),
+            )
+        })
+    }
+
+    async fn do_read(&mut self) -> io::Result<()> {
+        let n = tokio::time::timeout(Duration::from_secs(5), self.stream.read_buf(&mut self.buf))
```
Comment (pompon0) [inline chain/network/src/raw.rs:190]: hardcoding timeouts like this is generally not a good idea. How about you pass it as an argument? Ideally this future should be cancellable, but I guess we don't care that much about code used by probing.

File: chain/network/src/raw.rs [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+                    _ => return Err(e),
+                },
+            };
+        }
+        Ok(lengths)
+    }
+
+    // Reads from the socket until there is at least one full PeerMessage available.
+    // After that point, continues to read any bytes available to be read without blocking
+    // and appends any extra messages to the returned Vec. Usually there is only one in there
```
Comment (pompon0) [inline chain/network/src/raw.rs:272]: what's the purpose of this semantics?

File: chain/network/src/raw.rs [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+            self.do_read().await?;
+        }
+        Ok((self.extract_msg(msg_length)?, first_byte_time))
+    }
+
+    /// Reads from the socket until we receive some message that we care to pass to the caller
+    /// (that is, represented in `Message`). After receiving the first such message, we will
+    /// continue to read any messages available to be read without blocking. The `Instant`s
+    /// in the returned `Vec` are timestamps taken when the corresponding messages were read
+    pub async fn recv(&mut self) -> io::Result<Vec<(Message, Instant)>> {
```
Comment (pompon0) [inline chain/network/src/raw.rs:315]: Ok, I'm not exactly sure what's the goal, but it seems to me that this code is a bit overly complex. I've written sth like this to parse the message: https://github.com/near/nearcore/blob/06711f8460f946b8d2042aa1df6abe03c5184767/chain/network/src/peer/stream.rs#L155 and its essentially:
`
n = read.read_u32_le().await
let mut buf = vec![0; n];
read.read_exact(&mut buf[..]).await?;
`
where `read` is a TCP socket wrapped in a standard buffer of fixed size.

File: tools/ping/src/cli.rs [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+static CHAIN_INFO: &[ChainInfo] = &[
+    ChainInfo {
+        chain_id: "mainnet",
+        genesis_hash: CryptoHash([
+            198, 253, 249, 28, 142, 130, 248, 249, 23, 204, 25, 117, 233, 222, 28, 100, 190, 17,
+            137, 158, 50, 29, 253, 245, 254, 188, 251, 183, 49, 63, 20, 134,
+        ]),
+        head_height: 71112469,
```
Comment (pompon0) [inline tools/ping/src/cli.rs:97]: how is this value better than, for example, 0?

**Review by pompon0 — CHANGES_REQUESTED** [review]
Please add a unit test of the near_network part. You can find a plenty of examples in "chain/network/src/peer_manager/tests"

File: chain/network/src/raw.rs (hunk 2 above)
Comment (marcelo-gonzalez) [reply chain/network/src/raw.rs]: ok named it ReceivedMessage, since it's not always a Response (AccounceAccounts weren't requested by us)

File: chain/network/src/raw.rs (hunk 9 above)
Comment (marcelo-gonzalez) [reply chain/network/src/raw.rs]: inlined it

File: chain/network/src/raw.rs (hunk 11 above)
Comment (marcelo-gonzalez) [reply chain/network/src/raw.rs]: Ok so I did this originally because if you look at the trace level logs, you'll see that we do read more messages without blocking a lot of the time, and what I wanted to avoid was a situation where we get a big message, take a long time to deserialize it (in the case of a big `SyncRoutingTable`, 100ms+), and then report distorted latencies for subsequent pong messages. But looking at the logs some more, it's pretty clear that that only happens much at the beginning, and after that, we're talking about differences in reported latencies of < 1ms, so sure, I agree, let's just keep it simpler

File: chain/network/src/raw.rs (hunk 12 above)
Comment (marcelo-gonzalez) [reply chain/network/src/raw.rs]: made it simpler, although can't do `read_u32_le()` here, as then the future would not be cancellation-safe and not usable with `tokio::select`. actually did run into this a while ago when I was first working on this: https://github.com/near/nearcore/pull/7541/commits/ed5d20e5cc448e68988abb01f242f3f9fba30e53

File: tools/ping/src/cli.rs (hunk 13 above)
Comment (marcelo-gonzalez) [reply tools/ping/src/cli.rs]: true, no good reason

File: chain/network/src/raw.rs [hunk 14]
```diff
@@ -82,21 +87,32 @@ impl Peer {
         chain_id: &str,
         genesis_hash: CryptoHash,
         head_height: BlockHeight,
+        recv_timeout_seconds: u32,
```
Comment (pompon0) [inline chain/network/src/raw.rs:90]: use time::Duration rather than plain integer

File: chain/network/src/raw.rs [hunk 15]
```diff
@@ -343,3 +267,233 @@ impl Peer {
         Ok(())
     }
 }
+
+mod test {
+    use crate::network_protocol::{
+        Encoding, Handshake, PeerIdOrHash, PeerMessage, Pong, RawRoutedMessage, RoutedMessageBody,
+    };
+    use crate::raw;
+    use crate::time::Utc;
```
Comment (pompon0) [inline chain/network/src/raw.rs:276]: Don't embed test module into the implementation. Create separate files:
raw/mod.rs
raw/tests.rs

File: chain/network/src/raw.rs [hunk 16]
```diff
@@ -343,3 +267,233 @@ impl Peer {
...
+    use near_crypto::{KeyType, SecretKey};
+    use near_primitives::network::PeerId;
+    use std::io;
+    use tokio::io::{AsyncReadExt, AsyncWrite, AsyncWriteExt};
+    use tokio::net::TcpListener;
+
+    // we add #[allow(dead_code)] to everything here because for some reason this generates
+    // dead code warnings even though everything is very clearly used below... maybe it's this?
+    // https://github.com/rust-lang/rust/issues/46379
```
Comment (pompon0) [inline chain/network/src/raw.rs:285]: it happens because you haven't added a conditional compilation guard at the top of the module
It should be
```
#[cfg(test)]
mod test { ...}
```
but please move it to a separate file anyway.

File: chain/network/src/raw.rs [hunk 17]
```diff
@@ -343,3 +267,233 @@ impl Peer {
...
+            let peer_id = PeerId::new(secret_key.public_key());
+            let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
+            Self { secret_key, peer_id, listener }
+        }
+
+        async fn run(self, msg_sequence: &'static [PeerMsgType]) -> anyhow::Result<()> {
+            let (mut stream, _addr) = self.listener.accept().await.unwrap();
+            let (read, mut write) = stream.split();
+            let mut read = tokio::io::BufReader::new(read);
+
```
Comment (pompon0) [inline chain/network/src/raw.rs:395]: For your implementation we are interested in testing that it can communicate with the actual implementation of the PeerManager. Your current test implementation will not notice if some substantial change to the network protocol happened with requires changes to your ping tool. Please use crate::peer_manager::testonly::start to create a real instance of the PeerManager.

File: chain/network/src/raw.rs [hunk 18]
```diff
@@ -216,115 +213,42 @@ impl Peer {
...
-    // After that point, continues to read any bytes available to be read without blocking
-    // and appends any extra messages to the returned Vec. Usually there is only one in there
-    async fn recv_messages(&mut self) -> io::Result<Vec<(PeerMessage, Instant)>> {
-        let mut messages = Vec::new();
+    async fn recv_message(&mut self) -> io::Result<(PeerMessage, Instant)> {
         let (msg_length, first_byte_time) = self.read_msg_length().await?;
 
         while self.buf.remaining() < msg_length + 4 {
-            // TODO: measure time to last byte here
             self.do_read().await?;
```
Comment (pompon0) [inline chain/network/src/raw.rs:221]: why a loop, when you can just do the same as what you have written in the test below?:
```
 let len = read.read_u32_le().await.unwrap();
            let mut buf = vec![0; len as usize];
            read.read_exact(&mut buf).await.unwrap();
```
?

File: chain/network/src/raw.rs (hunk 14 above)
Comment (marcelo-gonzalez) [reply chain/network/src/raw.rs]: done

File: chain/network/src/raw.rs (hunk 16 above)
Comment (marcelo-gonzalez) [reply chain/network/src/raw.rs]: done

File: chain/network/src/raw.rs (hunk 17 above)
Comment (marcelo-gonzalez) [reply chain/network/src/raw.rs]: true, done

File: chain/network/src/raw.rs (hunk 18 above)
Comment (marcelo-gonzalez) [reply chain/network/src/raw.rs]: as I mentioned above, the pattern you suggest here would make `recv()` incorrect to use in a loop with `tokio::select` as is done in this PR, since the function can be stopped and restarted in between those two awaits. So then we could still try

```rust
        self.stream.read_exact(&mut self.buf[..4]).await?;
        let len = u32::from_le_bytes(self.buf[..4].try_into().unwrap());
        self.buf.reserve(len as usize);
        self.stream.read_exact(&mut self.buf[..len]).await?;
```

but we are not using a `tokio::io::BufReader`, so the initial read of 4 bytes is not great, and the current loop has us read a larger amount into the buffer at a time. When you unravel the above to get something more efficient without a `tokio::io::BufReader`, it's not a whole lot simpler than what's here.

so then we could definitely use a `tokio::io::BufReader` and use `read_exact()`. But then we have a similar issue where we need to keep the result around between any `await`s, so feels to me just much more direct to trade a little bit of simplicity for explicit control over when we `advance()` the buffer

File: chain/network/src/raw.rs (hunk 15 above)
Comment (marcelo-gonzalez) [reply chain/network/src/raw.rs]: done

File: chain/network/src/raw.rs (hunk 18 above)
Comment (pompon0) [reply chain/network/src/raw.rs]: Right, I missed your response there, sorry. So yes, you are right, to get cancellation safety you need a bit more that I suggested and it's fine if you want it implemented it this way.

File: chain/network/src/raw/tests.rs [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    let mut rng = testonly::make_rng(33955575545);
+    let rng = &mut rng;
+    let mut clock = time::FakeClock::default();
+    let chain = Arc::new(data::Chain::make(&mut clock, rng, 10));
+
+    let cfg = chain.make_config(rng);
+    let peer_id = cfg.node_id();
+    let addr = cfg.node_addr.unwrap();
+    let genesis_id = chain.genesis_id.clone();
+    let _pm = crate::peer_manager::testonly::start(
```
Comment (pompon0) [inline chain/network/src/raw/tests.rs:20]: much better, thx!

**Review by pompon0 — APPROVED** [review]
